onnxruntime/core/providers/trt/trt_execution_provider.cc

```diff
@@ -0,0 +1,420 @@
+    int num_outputs = output_binding_indexes.size();
+    ORT_ENFORCE(trt_engine->getNbBindings() == (num_inputs + num_outputs));
+
+    // Save engine, context and input/output info to map
+    parsers_.emplace(fused_node->Name(), std::move(trt_parser));
+    engines_.emplace(fused_node->Name(), std::move(trt_engine));
+    contexts_.emplace(fused_node->Name(), std::move(trt_context));
+    input_info_[fused_node->Name()].push_back(input_indexes);
+    input_info_[fused_node->Name()].push_back(input_binding_indexes);
+    input_info_[fused_node->Name()].push_back(input_dim_sizes);
+    output_info_[fused_node->Name()].push_back(output_binding_indexes);
+    output_info_[fused_node->Name()].push_back(output_dim_sizes);
+    output_shapes_[fused_node->Name()] = output_shapes;
+
+    // Create function state
+    NodeComputeInfo compute_info;
+    compute_info.create_state_func = [=](ComputeContext* context, FunctionState* state) {
+      std::unique_ptr<TRTFuncState> p = std::make_unique<TRTFuncState>();
+      *p = {context->allocate_func, context->release_func, context->allocator_handle, parsers_[context->node_name].get(), engines_[context->node_name].get(), contexts_[context->node_name].get(),
+            input_info_[context->node_name], output_info_[context->node_name], output_shapes_[context->node_name]};
+      *state = p.release();
+      return 0;
+    };
+
+    // Release function state
+    compute_info.release_state_func = [](FunctionState state) {
+      if (state)
+        delete static_cast<TRTFuncState*>(state);
+    };
+
+    // Create compute function
+    compute_info.compute_func = [](FunctionState state, ONNXRunTimeTensor* input_tensors, size_t num_inputs, ONNXRunTimeTensor* output_tensors, size_t num_outputs) {
+      ORT_UNUSED_PARAMETER(num_inputs);
+      ORT_UNUSED_PARAMETER(num_outputs);
+      TRTFuncState* trt_state = reinterpret_cast<TRTFuncState*>(state);
+      std::vector<int> input_indexes = (trt_state->input_info)[0];
+      std::vector<int> input_binding_indexes = (trt_state->input_info)[1];
+      std::vector<int> input_dim_sizes = (trt_state->input_info)[2];
+      std::vector<int> output_binding_indexes = (trt_state->output_info)[0];
+      std::vector<int> output_dim_sizes = (trt_state->output_info)[1];
+      std::vector<std::vector<int64_t>> output_shapes = trt_state->output_shapes;
+      int num_binding_inputs = input_binding_indexes.size();
+      int num_binding_outputs = output_binding_indexes.size();
+      cudaStream_t stream;
+      CHECK(cudaStreamCreate(&stream));
+      std::vector<void*> buffers(num_binding_inputs + num_binding_outputs);
+      std::vector<int> batch_size;
+
+      // Get batch size and allocate cuda memory for inputs
+      for (int i = 0, end = num_binding_inputs; i < end; ++i) {
+        const auto& tensor_input = input_tensors[input_indexes[i]];
+        const auto& tensor_shape = tensor_input.shape;
+        batch_size.push_back(tensor_shape[0]);
+
```

Comment:
There seems to be an assumption that the batch size for the first input will always be the largest, however that's very implicit.

If there should only be one batch size and all inputs should have the same value, can you instead validate that (and just use an int instead of a vector<int> for batch_size)?

Or if there can be a mix of batch sizes, but all values must be <= the batch size from the first input (given batch_size[0] is what's used in the memory allocation logic), can you validate that is the case?

Reply:
Removed the vector. and find largest input batch size as the batch size passed to TRT engine.

Reply:
Check if batch size is consistent. If not, throw error.